arrows/serialize/json/load_save_metadata.cxx

```diff
@@ -81,16 +81,23 @@ struct meta_item
     archive(  ::cereal::make_nvp( "type", item_value.type_name() ) );
     archive(  ::cereal::make_nvp( "name", trait.name() ) );
 
-    // This si a switch on the item data type
+    // This is a switch on the item data type
     if ( trait.is_floating_point() )
     {
       const double value = kwiver::vital::any_cast< double > ( this->item_value );
       archive( CEREAL_NVP( value ) );
     }
     else if ( trait.is_integral() )
     {
-      const uint64_t value = kwiver::vital::any_cast< uint64_t > ( this->item_value );
-      archive( CEREAL_NVP( value ) );
+      // bool metadata passes the is_integral() check but cannot be cast
+      // to uint64_t
+      if ( trait.tag_type() == typeid( bool ) ) {
+        const bool value = kwiver::vital::any_cast< bool > ( this->item_value );
+        archive( CEREAL_NVP( value ) );
+      } else {
+        const uint64_t value = kwiver::vital::any_cast< uint64_t > ( this->item_value );
```

Comment:
This may more of a question for @linus-sherrill, but do we need a special case for signed integers as well?  I expect this code may work as is, but the values of negative values may be written incorrectly in ASCII form in a JSON file since they will be cast to an unsigned int first.  We may even parse these value correctly if we read them back into an uint64 and then cast back to int, but I expect other JSON readers outside of KWIVER would not be so lucky.

Reply:
The original concept for the uint64 as the metadata type was just to hold 64 bits of untyped data and delegate extracting (casting) the actual data type to the customer (as opposed to adding support for {uint,unt}{8,16,32,64} types). A cast to the desired signed/unsigned type should reinterpret the bits to get the correct value. The bit pattern get serialized/deserialized and ends up the same at the other end. (JSON should encode this as type uint64_t.) The JSON generated by the save() method only has to be decoded by the load() method. The JSON format used here was only designed to be a data marshaling format used to move data between endpoints, but could be created by save() and parsed by something other than the load() method.  I think the JSON encoding is self consistent in that it states the type is uint64, even though it may be interpreted as a different type further up the stack. 

I don't know if it is burdensome to get the uint64_t from the metadata_item and them cast it to the expected integer type, but we could add an additional templated accessor method which would include the needed casting.

Reply:
@linus-sherrill in this case we are specifically looking for a JSON representation of metadata that can be exported outside of KWIVER and used by other readers.  I agree that for protobuf we can encode things however we want because our protobuf reader is the only thing reading it.  However, JSON is generally intended to be both human readable and more portable across tools.  Can we updated the JSON reader to correctly read and write signed types? or does this JSON serializer need to be consistent with protobuf?

Note that I don't think we need to encode all combinations of bit depths for integers as long as we cast up to the bigger bit depth.  It's really just a signed/unsigned issue.  Negative numbers will not cast to meaningful values as uint64.  So maybe we need both uint64 and int64.

Reply:
The JSON serializer does not need to be consistent with the protobuf format. The only hard requirement with the serializers is that deserializing serialized data returns the original value. I think that 

    int i = -1;
    unit64_t ui = static_cast<uint64_t>(i);
    // The unsigned representation for -1 is a really big number (0xFFFFFFFFFFFFFFFF)
    int j = static_cast<int>(ui);   // j should be -1
    int16_t k = static_cast<int16_t>(ui); // k should be -1

So it should be ok for the consumer to cast the result into the desired integer type with a static cast.

When converting signed to unsigned or a larger size integer, the high order sign bit is propagated through the additional high order bits. When converting to a integer of smaller size, the high order bits are truncated. This is not the case when converting floating point. I'm not sure if it works this way in python.

We could add a int64_t case to the serializer, although the only integer format that is currently used in the metadata is uint64_t. It is only by convention that a very limited set of types are used in the metadata.